### PR TITLE
Feat/2271 rust oxidationlibparseccratesplatform async crate mvp

### PR DESCRIPTION
**Pull request by FirelightFlagboy**

### CODE REVIEW

File: newsfragments/2223.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fixed clicking on a file link in some instances
```
Comment (touilleMan) [inline newsfragments/2223.bugfix.rst:1]: This newsfragments don't seem related to this PR...

Comment (FirelightFlagboy) [reply newsfragments/2223.bugfix.rst:1]: hum, I'll remove them

**Comment by FirelightFlagboy** [comment]
wololo, I've made some cleanup @touilleMan

File: oxidation/crates/platform-async/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+[package]
+name = "platform-async"
```
Comment (touilleMan) [inline oxidation/crates/platform-async/Cargo.toml:2]: `libparsec_platform_async` (the idea is it's easy to see the correspondance with the path `oxidation/libparsec/crates/platform_async` while still behind explicit about the fact this is not a 3rd party crate)

And we use underscore instead of dash (because it's what is used in the rust code when doing `use libparsec_platform_async::*`, so better stay consistent)

File: oxidation/crates/platform-async/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+[dependencies]
+flume = "0.10.12"
+futures = "0.3.21"
+futures-lite = { version = "1.12.0", default-features = false }
+
+[dev-dependencies]
+log = "0.4.17"
+
+[target.'cfg(target_arch = "wasm32")'.dependencies]
```
Comment (touilleMan) [inline oxidation/crates/platform-async/Cargo.toml:14]: so far we use feature `web` vs `native` instead of relying on target_arch (given `wasm32` may refer to wasi, or to a node module)

I guess the advantage of the target_arch, though less consistent, is to avoid having to deal with feature and default_features in the crates but I have the feeling this is not a big deal given we have to set the crates once and for all with default_features=false for the dependencies and we're good ;-)

File: oxidation/crates/platform-async/src/task.rs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BSLv1.1 (eventually AGPLv3) 2016-2021 Scille SAS
+
+use std::future::Future;
+
+pub trait Taskable<T>: Future<Output = T> {
```
Comment (touilleMan) [inline oxidation/crates/platform-async/src/task.rs]: What's the need for this trait ?

I guess this is to abstract the fact we have 2 implementations of Task (desktop vs web). If this is the fact I think we are better without any trait at all.
This is the approach used for the crate `parsec_api_crypto` (that have rustcrypto and libsodium implementations).

This has be benefit of simplifying the code (no need to use the trait in the caller code, simpler types to work with, async traits add complexity)
The only drawback is it makes impossible to use both implementations at the same time, but this is not a concern in here ^^

File: oxidation/crates/platform-async/Cargo.toml (hunk 2 above)
Comment (FirelightFlagboy) [reply oxidation/crates/platform-async/Cargo.toml:2]: it don't have `libparsec` in the path :p 

for the crate name, I have a lot of counter example that use the kebab case for it

File: oxidation/crates/platform-async/src/platform.rs [hunk 5]
```diff
@@ -0,0 +1,8 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BSLv1.1 (eventually AGPLv3) 2016-2021 Scille SAS
```
Comment (touilleMan) [inline oxidation/crates/platform-async/src/platform.rs]: we use `<module_name>/mod.rs` style in the codebase

File: oxidation/crates/platform-async/src/platform/default.rs [hunk 6]
```diff
@@ -0,0 +1,149 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BSLv1.1 (eventually AGPLv3) 2016-2021 Scille SAS
```
Comment (touilleMan) [inline oxidation/crates/platform-async/src/platform/default.rs]: maybe call it "native.rs" or "tokio.rs" instead ?

File: oxidation/crates/platform-async/src/platform/default.rs [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    F: Future + Send + 'static,
+    F::Output: Send + 'static,
+{
+    let shared_state = SharedState::default();
+    let shared_state = Arc::new(Mutex::new(shared_state));
+
+    let task = Task::new(shared_state.clone());
+    let runnable = Runnable::new(future, shared_state);
+
+    let _handle = tokio::spawn(runnable);
```
Comment (touilleMan) [inline oxidation/crates/platform-async/src/platform/default.rs]: would be useful in this kind of code to explain why dropping is okay

(btw, is there a rustonic way of showing drop without using the variable is not an error in the code ? using `drop(handle)` explicitly ?)

File: oxidation/crates/platform-async/Cargo.toml (hunk 3 above)
Comment (FirelightFlagboy) [reply oxidation/crates/platform-async/Cargo.toml:14]: I can use `target_family = "wasm"`. 

I prefer `target_*` over `feature = web / native` because it's more explicite that the architecture/capabilities will be different

File: oxidation/crates/platform-async/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,26 @@
+[package]
```
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply oxidation/crates/platform-async/Cargo.toml]: I've moved the crate

File: oxidation/crates/platform-async/src/task.rs [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+    /// Cancels the task.
+    fn cancel(&self) -> Option<T>;
+
+    /// Detaches the task to let it keep running in the backgroud.
+    fn detach(self);
+
+    /// Return `true` if the current task is finished.
+    fn is_finised(&self) -> bool;
```
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply oxidation/crates/platform-async/src/task.rs]: done

**Review by touilleMan — CHANGES_REQUESTED** [review]

File: oxidation/crates/platform-async/src/task.rs (hunk 4 above)
Comment (FirelightFlagboy) [reply oxidation/crates/platform-async/src/task.rs]: It's to make a sure that any implementations will share a set of method

File: oxidation/crates/platform-async/src/platform/default.rs (hunk 7 above)
Comment (FirelightFlagboy) [reply oxidation/crates/platform-async/src/platform/default.rs]: clippy may don't like `drop(object)` https://rust-lang.github.io/rust-clippy/master/#drop_copy

File: oxidation/crates/platform-async/examples/eyeballs.rs [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+    let task = spawn(attempt(
+        which,
+        targets,
+        failed_attempt,
+        join_set.clone(),
+        sender,
+    ));
+
+    join_set.lock().unwrap().push(task);
```
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply oxidation/crates/platform-async/examples/eyeballs.rs]: it's an example ;) but I take you comment into consideration when I'll be implementing the joinset

File: oxidation/crates/platform-async/src/platform.rs (hunk 5 above)
Comment (FirelightFlagboy) [reply oxidation/crates/platform-async/src/platform.rs]: I've moved the file.

I've looked if we can configure clippy for the whole project:

It's not possible at the moment but https://github.com/rust-lang/cargo/issues/5034

File: oxidation/libparsec/crates/platform-async/examples/eyeballs.rs [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+#[derive(Clone, Copy, Debug)]
+enum Mode {
+    Fail(Duration),
+    Success(Duration),
+}
+
+#[tokio::main]
+async fn main() -> Result<(), ()> {
+    simple_logger::init_with_level(log::Level::Debug).expect("cannot initialise simple logger");
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/platform-async/examples/eyeballs.rs:24]: initialize

File: oxidation/libparsec/crates/platform-async/examples/eyeballs.rs [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+    let join_set = Vec::with_capacity(targets.len());
+    let join_set = Arc::new(Mutex::new(join_set));
+    let (tx, mut rx) = mpsc::channel::<usize>(targets.len());
+
+    spawn_attempt(0, targets, failed_attempt, join_set.clone(), tx);
+    let result = dbg!(rx.recv().await);
+    for task in join_set.lock().unwrap().iter() {
+        task.cancel();
+    }
+    result.ok_or(())
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/platform-async/examples/eyeballs.rs:57]: What's the point to transform the option ?

File: oxidation/libparsec/crates/platform-async/examples/eyeballs.rs [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+            failed_attempt[which].notify_waiters();
+        }
+        Ok(_) => {
+            log::info!("[#{which}] finished with success");
+            sender.send(which).await.unwrap();
+        }
+    }
+}
+
+async fn mock_load(mode: Mode) -> Result<(), ()> {
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/platform-async/examples/eyeballs.rs:126]: Better to use an Enum as Result to describe more clearly

File: oxidation/libparsec/crates/platform-async/src/platform/default.rs [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+    F: Future + Send + 'static,
+    F::Output: Send + 'static,
+{
+    let shared_state = SharedState::default();
+    let shared_state = Arc::new(Mutex::new(shared_state));
+
+    let task = Task::new(shared_state.clone());
+    let runnable = Runnable::new(future, shared_state);
+
+    let _handle = tokio::spawn(runnable);
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/platform-async/src/platform/default.rs:23]: unused variable

**Review by TimeEngineer — CHANGES_REQUESTED** [review]
Some minor change

File: oxidation/libparsec/crates/platform-async/examples/eyeballs.rs (hunk 11 above)
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/platform-async/examples/eyeballs.rs]: corrected in 001824770

File: oxidation/libparsec/crates/platform-async/examples/eyeballs.rs [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+        .iter()
+        .map(|_| Arc::new(Notify::new()))
+        .collect::<Vec<_>>();
+    let targets = Arc::new(targets);
+    let join_set = Vec::with_capacity(targets.len());
+    let join_set = Arc::new(Mutex::new(join_set));
+    let (tx, mut rx) = mpsc::channel::<usize>(targets.len());
+
+    spawn_attempt(0, targets, failed_attempt, join_set.clone(), tx);
+    let result = dbg!(rx.recv().await);
```
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/platform-async/examples/eyeballs.rs:53]: it's an example ;)

File: oxidation/libparsec/crates/platform-async/examples/eyeballs.rs (hunk 13 above)
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/platform-async/examples/eyeballs.rs:126]: `Result` is an `Enum`, what would you prefer ?

File: oxidation/libparsec/crates/platform-async/examples/eyeballs.rs (hunk 12 above)
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/platform-async/examples/eyeballs.rs:57]: none

File: oxidation/libparsec/crates/platform-async/src/platform/default.rs [hunk 16]
```diff
@@ -20,7 +20,7 @@ where
     let task = Task::new(shared_state.clone());
     let runnable = Runnable::new(future, shared_state);
 
-    let _handle = tokio::spawn(runnable);
+    drop(tokio::spawn(runnable));
```
Comment (TimeEngineer) [inline oxidation/libparsec/crates/platform-async/src/platform/default.rs:23]: why drop ?
It reaches out of scope after the return

File: oxidation/libparsec/crates/platform-async/examples/eyeballs.rs (hunk 12 above)
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/platform-async/examples/eyeballs.rs]: changed in bb524deee

**Review by TimeEngineer — CHANGES_REQUESTED** [review]

File: oxidation/libparsec/crates/platform-async/examples/eyeballs.rs (hunk 13 above)
Comment (TimeEngineer) [reply oxidation/libparsec/crates/platform-async/examples/eyeballs.rs:126]:
```Rust
MockResult {
    Success,
    Failed,
}
```
Result is an Enum, but generic Enum, why should I want to think about if there is something wrapped inside when I read a match arms ?
It's like why should I use Enum instead of bool ^^

Comment (TimeEngineer) [reply oxidation/libparsec/crates/platform-async/examples/eyeballs.rs:126]: You can also keep your Mode Enum
Do you know this pattern ?
```Rust
enum Enum {
    Foo(i32),
    Bar(i32),
}

fn fun() -> Enum {
    let e = Enum::Foo(5);
    let (Enum::Foo(delay) | Enum::Bar(delay)) = e;
    // sleep delay
    e
}
```

Comment (TimeEngineer) [reply oxidation/libparsec/crates/platform-async/examples/eyeballs.rs:126]: I'm wondering what's the best way for this type of structure
```rust
struct Foo {
    delay: i32
    mode: Mode
}
enum Mode {
    Success,
    Fail,
```
or the current implementation

Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/platform-async/examples/eyeballs.rs:126]:
```rust
let e = Enum::Foo(5);
let (Enum::Foo(delay) | Enum::Bar(delay)) = e;
```
I know this pattern

Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/platform-async/examples/eyeballs.rs:126]: > I'm wondering what's the best way for this type of structure
> 
> ```rust
> struct Foo {
>     delay: i32,
>     mode: Mode,
> }
> enum Mode {
>     Success,
>     Fail,
> }
> ```
> 
> or the current implementation

I like this structure, but keep in mind that this is an example

File: oxidation/libparsec/crates/platform-async/src/platform/native/task.rs [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+    detached: bool,
+    value: Option<T>,
+    task_waker: Option<Waker>,
+    runnable_waker: Option<Waker>,
+}
+
+impl<T> Default for SharedState<T> {
+    fn default() -> Self {
+        Self {
+            canceled: false,
```
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/platform-async/src/platform/native/task.rs:40]: yes it can have multiple combination

File: oxidation/libparsec/crates/platform-async/src/platform/default.rs (hunk 16 above)
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/platform-async/src/platform/default.rs]: I don't need the handle returned by tokio.

I've removed the drop in 847d250

File: oxidation/libparsec/crates/platform-async/src/native/timer.rs [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+            sleep: Box::pin(sleep(duration)),
+        }
+    }
+}
+
+impl Future for Timer {
+    type Output = ();
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.sleep.poll_unpin(cx)
```
Comment (touilleMan) [inline oxidation/libparsec/crates/platform-async/src/native/timer.rs:42]: Use `Pin::new(self).poll(cx)` to avoid needed future ? (cf. https://rust-lang-nursery.github.io/futures-api-docs/0.1.27/src/futures_util/future/mod.rs.html#531)

File: oxidation/libparsec/crates/platform-async/src/native/task.rs [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+        state.canceled = true;
+        if let Some(ref waker) = state.runnable_waker {
+            waker.wake_by_ref();
+        }
+        state.value.take()
+    }
+
+    /// Detaches the task to let it keep running in the background
+    pub fn detach(self) {
+        self.shared_state.lock().unwrap().detached = true;
```
Comment (touilleMan) [inline oxidation/libparsec/crates/platform-async/src/native/task.rs:87]: expect instead of unwrap ?

File: oxidation/libparsec/crates/platform-async/src/native/task.rs [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+    shared_state: Arc<Mutex<SharedState<T>>>,
+}
+
+impl<T> Task<T> {
+    fn new(shared_state: Arc<Mutex<SharedState<T>>>) -> Self {
+        Self { shared_state }
+    }
+
+    /// Cancels the task
+    pub fn cancel(&self) -> Option<T> {
```
Comment (touilleMan) [inline oxidation/libparsec/crates/platform-async/src/native/task.rs:75]: cancel is too close to what we use in trio, we should use another more explicit name (to show the fact the future is not longer going to be called anymore)

so maybe `abort` to use the tokio jargon

File: oxidation/libparsec/crates/platform-async/src/native/timer.rs (hunk 18 above)
Comment (TimeEngineer) [reply oxidation/libparsec/crates/platform-async/src/native/timer.rs:42]: I prefer ctx instead of cx, because it is used in a lot of crates

Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/platform-async/src/native/timer.rs:42]: `cx` is the default name of the parameters in the trait definition

Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/platform-async/src/native/timer.rs:42]: https://doc.rust-lang.org/std/future/trait.Future.html

Comment (TimeEngineer) [reply oxidation/libparsec/crates/platform-async/src/native/timer.rs:42]: God :man_facepalming:

File: oxidation/libparsec/crates/platform-async/examples/eyeballs.rs [hunk 21]
```diff
@@ -0,0 +1,134 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BSLv1.1 (eventually AGPLv3) 2016-2021 Scille SAS
+#![cfg(not(target_arch = "wasm32"))]
```
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/platform-async/examples/eyeballs.rs:2]: done in 77c6eba35

File: oxidation/libparsec/crates/platform-async/src/native/task.rs (hunk 19 above)
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/platform-async/src/native/task.rs]: done in 9de5859

**Comment by FirelightFlagboy** [comment]
@touilleMan and @TimeEngineer

Could you review the changes that I've made

**Review by TimeEngineer — APPROVED** [review]
